modules/hivemq-edge-module-modbus/src/main/java/com/hivemq/edge/adapters/modbus/impl/ModbusClient.java

```diff
@@ -18,160 +18,133 @@
-            throw new ProtocolAdapterException(e);
-        }
+    public CompletableFuture<Void> connect() {
+        return modbusClient.connect().thenApply(unused -> null);
     }
 
-    public @NotNull DataPoint readHoldingRegisters(final int startIdx, final int count, final @NotNull ModbusDataType dataType)
-            throws ProtocolAdapterException {
-        try {
-            final ModbusTcpMaster client = getOrCreateClient();
-            final CompletableFuture<ReadHoldingRegistersResponse> future =
-                    client.sendRequest(new ReadHoldingRegistersRequest(startIdx, Math.min(count, 125)), 0);
-            return future.thenApply(response -> {
+    /**
+     * Coils are 1bit.
+     */
+    public @NotNull CompletableFuture<DataPoint> readCoils(final int startIdx, final int count, final int unitId) {
+        return modbusClient
+            .<ReadCoilsResponse>sendRequest(new ReadCoilsRequest(startIdx, Math.min(count, 2000)), unitId)
+            .thenApply(response -> {
                 try {
-                    final ByteBuf buf = response.getRegisters();
+                    final ByteBuf buf = response.getCoilStatus();
                     return dataPointFactory.create("registers-" + startIdx + "-" + (startIdx + count),
-                            convert(buf, dataType, count));
+                            convert(buf, ModbusDataType.BOOL, count));
                 } finally {
                     ReferenceCountUtil.release(response);
                 }
-            }).get();
+            });
+    }
 
-        } catch (final Exception e) {
-            throw new ProtocolAdapterException(e);
-        }
+    /**
+     * Discrete registers are 1bit.
+     */
+    public @NotNull CompletableFuture<DataPoint> readDiscreteInput(final int startIdx, final int count, final int unitId) {
+        return modbusClient
+                .<ReadDiscreteInputsResponse>sendRequest(new ReadDiscreteInputsRequest(startIdx, Math.min(count, 2000)), unitId)
+                .thenApply(response -> {
+                    try {
+                        final ByteBuf buf = response.getInputStatus();
+                        return dataPointFactory.create("registers-" + startIdx + "-" + (startIdx + count),
+                                convert(buf, ModbusDataType.BOOL, count));
+                    } finally {
+                        ReferenceCountUtil.release(response);
+                    }
+                });
     }
 
-    public Short[] readInputRegisters(final int startIdx, final int count) throws ProtocolAdapterException {
+    /**
+     * Holding registers are 16bit.
+     */
+    public @NotNull CompletableFuture<DataPoint> readHoldingRegisters(final int startIdx, final int count, final @NotNull ModbusDataType dataType, final int unitId) {
+        return modbusClient
+                .<ReadHoldingRegistersResponse>sendRequest(new ReadHoldingRegistersRequest(startIdx, Math.min(count, 125)), unitId)
```

Comment:
what is this magical number 125? Pls extract and give it a good naming

Reply:
Constants from the modbus spec.
I extracted them into constants.